### PR TITLE
Troubleshooting design doc

### PR DESCRIPTION
**Pull request by IanLewis**
The purpose of this design doc is to improve the troubleshooting and debugging features of the Dashboard UI. The main addition is a feature to open a terminal session in a container (#1345)

### CODE REVIEW

File: docs/design/troubleshooting.md [hunk 1]
```diff
@@ -0,0 +1,119 @@
+# Application Troubleshooting 
+
+This design document is a proposal to consolidate add simple application
```
Comment (bryk) [inline docs/design/troubleshooting.md]: "consolidate add"? Typo?

File: docs/design/troubleshooting.md [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+The UI will be improved to add tabs to a pod detail page. The tabs will include:
+
+- Pod details
+- Logs
+- Terminal
+
+The tabs will allow the user to switch between the tabs without ending the
+session in another tab (i.e. without reloading the page). This will be
+especially important for the terminal feature so that users can view logs or
+detail information about the pod while working and without loosing their place.
```
Comment (danielromlein) [inline docs/design/troubleshooting.md]: >loosing

losing

File: docs/design/troubleshooting.md [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+Troubleshooting and debugging have consistently been a top priority from a
+users point of view. See:
+
+- [Kubernetes UI Survey Nov 2016](https://github.com/kubernetes/dashboard/blob/master/docs/devel/images/kubecon-dashboard-survey-infographic-nov2016.png)
+- [Kubernetes UI Survey Nov 2017](http://blog.kubernetes.io/2017/01/kubernetes-ux-survey-infographic.html)
+
+## Problem Statement
+
+- Update the UI to allow easier troubleshooting of problem containers
+  - Allow users to execute process in containers running in pods and open a
```
Comment (danielromlein) [inline docs/design/troubleshooting.md]: >execute process

"execute a process" or "execute processes"

File: docs/design/troubleshooting.md [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+terminal session is established. The terminal will attempt to run the standard
+'/bin/sh' shell command in the container. If this command is present and
+executable it should allow end users to execute arbitrary commands in the shell
+environment.
+
+![](mockups/24-01-2017-troubleshooting/terminal.png)
+
+The terminal will allow the user to switch between containers in the pod and
+will show the status of the session. If the user's session becomes disconnected,
+the terminal application will incidate this and allow the user to reconnect.
```
Comment (danielromlein) [inline docs/design/troubleshooting.md]: >incidate

indicate

**Comment by danielromlein** [comment]
Thanks for putting this together @IanLewis!

  - Could you please expound on _why_ a user would take the action of opening a terminal session?

- Currently the "_Details_" view of a pod contains the sections _Details_, _Containers_, _Conditions_, and _Created by_ – where do you foresee these cards going in this information architecture? What alternative nomenclature could we employ so "Details" doesn't appear twice?

**Comment by IanLewis** [comment]
@danielromlein Thanks for the review!

1. There are so many things one might want to do in a terminal to list them all but I'll try to add some typical use cases.
2. When I say "Details" I mean all of the items shown on the current pod details page. I guess I considered all of those things the pod's "details". I'll update the doc that to make it clear.

File: docs/design/troubleshooting.md (hunk 4 above)
Comment (IanLewis) [inline docs/design/troubleshooting.md]: Done.

File: docs/design/troubleshooting.md (hunk 3 above)
Comment (IanLewis) [inline docs/design/troubleshooting.md]: Done.

File: docs/design/troubleshooting.md (hunk 2 above)
Comment (IanLewis) [inline docs/design/troubleshooting.md]: Done.

File: docs/design/troubleshooting.md (hunk 1 above)
Comment (IanLewis) [inline docs/design/troubleshooting.md]: Done

**Comment by IanLewis** [comment]
@danielromlein

>   - Could you please expound on _why_ a user would take the action of opening a terminal session?

I added a new "Use Cases" section that hopefully answer your questions. PTAL

> - Currently the "_Details_" view of a pod contains the sections _Details_, _Containers_, _Conditions_, and _Created by_ – where do you foresee these cards going in this information architecture? What alternative nomenclature could we employ so "Details" doesn't appear twice?

I updated the doc to indicate where the cards would go. I guess I see the current cards as all being "Details" of the Pod. The current details lists basic details, so maybe "Basic Details" would work? Or "Basic Information"? "Properties"?  @bryk WDYT?

**Comment by bryk** [comment]
> I updated the doc to indicate where the cards would go. I guess I see the current cards as all being "Details" of the Pod. The current details lists basic details, so maybe "Basic Details" would work? Or "Basic Information"? "Properties"? @bryk WDYT?

I'd stay with "Details" or "Overview" as our plan is to put more information there.

**Comment by IanLewis** [comment]
Ok. Let's call the tab "Overview" and keep the name "Details" for the card name. How's that?

**Comment by danielromlein** [comment]
@IanLewis Thanks for the use case info! I think that's super valuable to have documented.
Just to make sure I'm understanding correctly: the advantage of this to users (over just switching to Terminal) is that they're already in the location that they want to execute commands regarding / view logs for...?

>Ok. Let's call the tab "Overview" and keep the name "Details" for the card name. How's that?

Sounds good to me! This also relates to #1248 and the tabular structure we were exploring for that. Naming for #1248 was left unfinalized, but I think starting with _Overview_, _Logs_, and _Terminal_ for this is a good place to start. Eventually we can add the _Meta_ and _Relationships_ tabs. 

Just to be clear: we're not planning on incorporating the pop-out functionality initially, correct?

**Comment by IanLewis** [comment]
@danielromlein 

> @IanLewis Thanks for the use case info! I think that's super valuable to have documented.
Just to make sure I'm understanding correctly: the advantage of this to users (over just switching to Terminal) is that they're already in the location that they want to execute commands regarding / view logs for...?

If I get what you are saying, yes. The idea is that a user might be looking at the Dashboard to figure out what's wrong and they see something in the pod details, or logs that makes them want to enter the container and execute commands to debug more. They may also want to switch back to look at the details or logs to verify what they are seeing etc.

> Sounds good to me! This also relates to #1248 and the tabular structure we were exploring for that. Naming for #1248 was left unfinalized, but I think starting with _Overview_, _Logs_, and _Terminal_ for this is a good place to start. Eventually we can add the _Meta_ and _Relationships_ tabs. 

Ok. I didn't realize you were planning on breaking out those parts as well. Maybe you're right and this is the best way to go.

> Just to be clear: we're not planning on incorporating the pop-out functionality initially, correct?

Yah, I think that keeping it to a tab that can be switched back and forth is ok for now at least. I have some ideas of what a popout might look like but I'd like to leave that until later.

**Comment by danielromlein** [comment]
@IanLewis could you send me a link to a working instance of this that I could check out? I'm not familiar yet with the running locally workflow.

**Comment by IanLewis** [comment]
@danromlein This PR just has docs and Balsamiq mockup. I haven't started on
implementing it in Dashboard yet. For this PR I just want to make sure the
design looks ok to everyone before implementing.

On Thu, Feb 9, 2017, 00:35 Dan Romlein <notifications@github.com> wrote:

@IanLewis <https://github.com/IanLewis> could you send me a link to a
working instance of this that I could check out? I'm not familiar yet with
the running locally workflow.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/dashboard/pull/1585#issuecomment-278362146>,
or mute the thread

**Comment by danielromlein** [comment]
Ah, OK got it. Do we think having a higher-fidelity mockup would be beneficial or are you planning on implementing just from these wireframes? I'm available to mock that up in Sketch.

**Comment by IanLewis** [comment]
@danromlein No I think we are good. If we are cool with this design moving forward I'll merge this and start implementing and create a WIP PR for the initial changes.

**Pull request merged**